File: src/main/java/com/github/dwendelen/platformd/core/budget/ExpensePost.java
package com.github.dwendelen.platformd.core.budget;

import com.github.dwendelen.platformd.core.budget.command.CreateExpensePost;
import com.github.dwendelen.platformd.core.budget.command.CreateIncomeSource;
import com.github.dwendelen.platformd.core.budget.event.ExpensePostCreated;
import com.github.dwendelen.platformd.core.budget.event.IncomeSourceCreated;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.commandhandling.model.AggregateIdentifier;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.spring.stereotype.Aggregate;

import java.util.UUID;

import static org.axonframework.commandhandling.model.AggregateLifecycle.apply;

@Aggregate
public class ExpensePost {
    @AggregateIdentifier
    private UUID uuid;

    @CommandHandler
    public ExpensePost(CreateExpensePost command) {
        apply(new ExpensePostCreated()
            .setUuid(command.getUuid())
            .setName(command.getName())
        );
    }

    @EventSourcingHandler
    public void on(ExpensePostCreated event) {
        this.uuid = event.getUuid();
    }
}

File: application/src/main/scala/com/github/dwendelen/platformd/core/budget/ExpensePost.scala
package com.github.dwendelen.platformd.core.budget

import java.util.UUID

import com.fasterxml.jackson.annotation.JsonCreator
import com.github.dwendelen.platformd.infrastructure.cqrs.{Aggregate, AggregateResult, CommandSucces}

class ExpensePost(uuid: UUID) extends Aggregate[ExpensePost] {
    def this() = this(null)

    override def applyEvent(event: Any): ExpensePost = event match {
        case e: ExpensePostCreated => applyEvent(e)
        case _ => this
    }

    def applyEvent(event: ExpensePostCreated): ExpensePost =
        new ExpensePost(event.uuid)

    override def handle(command: Any): AggregateResult = command match {
        case cmd: CreateExpensePost => AggregateResult(CommandSucces(cmd.uuid),handle(cmd))
        case cmd: RecordExpense => AggregateResult(CommandSucces(uuid), handle(cmd))
    }
    def handle(cmd: CreateExpensePost): List[Any] =
        List(new ExpensePostCreated(
            cmd.uuid,
            cmd.owner,
            cmd.name
        ))

    def handle(cmd: RecordExpense): List[Any] =
        List(new ExpenseRecorded(
            cmd.expensePost,
            cmd.transactionId,
            cmd.amount,
            cmd.accountId
        ))
}

//Events
@JsonCreator
class ExpensePostCreated(val uuid: UUID,
                          val owner: UUID,
                          val name: String)

@JsonCreator
class ExpenseRecorded(val expensePost: UUID,
                     val transactionId: UUID,
                     val amount: BigDecimal,
                     val accountId: UUID)

//Commands
@JsonCreator
class CreateExpensePost(val  uuid: UUID,
                         val owner: UUID,
                         val name: String)

@JsonCreator
class RecordExpense(val expensePost: UUID,
                   val transactionId: UUID,
                   val amount: BigDecimal,
                   val accountId: UUID)